Go template and init stage update

nikandfor opened the pull request:
* moved main package from ./app to ./cmd/mygit
* updated your_git.sh
* added go.mod

nikandfor commented:
I've committed with that commit applied so I need it in CI pipeline to be successful here.

https://github.com/codecrafters-io/course-sdk/pull/3

rohitpaulk commented on starter_templates/go/your_git.sh:
```diff
@@ -6,6 +6,10 @@
+
+( cd $(dirname "$0") && go build -o "$tmpFile" ./cmd/mygit )
+
+"$tmpFile" "$@"
```
Let's keep the exec call as before - we've seen problems (can't remember what exactly atm, but git history might have details) when not using it

nikandfor replied:
git history starts from a big "copy from docker" commit where exec already exists.
I've removed it to make trap working. Otherwise temporary files kept building up.
But this is not so important, I'll return it back.

rohitpaulk replied:
Ah, got it - yeah, figured this might be related to `trap`. 

Here's the old discussion I was referring to re: why exec is needed: https://github.com/codecrafters-io/languages/pull/31#discussion_r446638753.

nikandfor replied:
Alternative, although it's quite more complex
https://unix.stackexchange.com/questions/146756/forward-sigterm-to-child-in-bash

rohitpaulk replied:
Nice find! I think something like that would be a good choice if this was internal and not exposed to users - but since this appears in the starter code, it could pretty confusing for users 🙂

rohitpaulk commented on starter_templates/go/go.mod:
```diff
@@ -0,0 +1,3 @@
+module git.codecrafters.io/go-git
+
+go 1.16
```
Should this be 1.19?

nikandfor replied:
No I've set the smallest supported version. This will make newer language features to not to compile even if using new compiler. But this doesn't work for stdlib changes.

nikandfor replied:
Nope, the compiler will complain about it. But our older version tests will work in return.

rohitpaulk replied:
Ah, okay got it. Sounds reasonable for now! I suppose at some point in the future it might make sense to deprecate 1.16 like we did with 1.13.

rohitpaulk commented on starter_templates/go/go.mod:
```diff
@@ -0,0 +1,3 @@
+module git.codecrafters.io/go-git
```
Assuming this has no bearing on functionality, let's make this line: 

```suggestion
module github.com/codecrafters-io/git-starter-go
```

rohitpaulk commented on starter_templates/go/cmd/mygit/main.go:
```diff
@@ -0,0 +1,35 @@
+import (
+	"fmt"
+	// Uncomment this block to pass the first stage!
+	//	"os"
```
```suggestion
	// "os"
```

nikandfor replied:
Uncomment script doesn't reformat the go code according to gofmt standard and as a result we get incorrectly formatted starter code. There are spaces instead of tabs and closing bracket appears in the wrong column.

As I understood uncomment script just removes slashes and spaces before them. So having all the needed tabs after slashes makes init stage correctly formatted.

nikandfor replied:
The pattern includes leading spaces
https://github.com/codecrafters-io/course-sdk/blob/7ba30ee84dd028d0c95222e06f22a89ff126fb96/lib/uncommenter.rb#L3

nikandfor replied:
Ah, there is grouping, got it

rohitpaulk replied:
(Sorry, replied before I saw your comment 🙂)

nikandfor replied:
Thanks for the explanation anyway.

I found why there was a problem. It was bad formatted in a starter template.

https://github.com/codecrafters-io/build-your-own-git/blob/0bdc768f45133e5ca4aaf58b4d081064c61f7af3/starter_templates/go/app/main.go#L28

rohitpaulk commented on starter_templates/go/cmd/mygit/main.go:
```diff
@@ -0,0 +1,35 @@
+
+	// Uncomment this block to pass the first stage!
+	//
+	//	switch command := os.Args[1]; command {
```
```suggestion
	// switch command := os.Args[1]; command {
```

(similar changes for other commented lines in this file. The whole "block" should be indented by `//` followed by a single space).

rohitpaulk commented on solutions/go/read_blob/diff/your_git.sh.diff:
```diff
@@ -0,0 +1,18 @@
+-( cd $(dirname "$0") && go build -o "$tmpFile" ./cmd/mygit )
+-
+-"$tmpFile" "$@"
++go build -o "$tmpFile" $(dirname "$0")/app/*.go
```
@nikandfor let's review all these diffs carefully and ensure they are good for production - I'd expect changes from stage 1 to be propagated to all `your_git.sh` files so that a diff is never shown

nikandfor replied:
Yes, this is the reason I wanted to remove the following stages and add them back merging with previous step code manually.

But I can propagate changes either.

rohitpaulk commented on solutions/go/read_blob/diff/go.mod.diff:
```diff
@@ -0,0 +1,4 @@
+@@ -1,3 +0,0 @@
+-module git.codecrafters.io/go-git
```
We don't want this showing up as a diff - the diffs are the only content users see in the app, and they must only contain changes relevant to the particular stage

nikandfor replied:
I've only made the CI not complain about the latter stages. I was planning to update and commit stages in different PRs.

nikandfor replied:
This was about init stage only.

rohitpaulk commented on solutions/go/read_blob/diff/cmd/mygit/main.go.diff:
```diff
@@ -0,0 +1,30 @@
+@@ -1,29 +0,0 @@
+-package main
+-
```
Same here - let's make changes to the other stages so that the diffs are appropriate. It's okay to not "cleanup" code from the future stages (that can be left for future PRs), but we need the diffs to be presentable to users

nikandfor replied:
ok

rohitpaulk reviewed (changes requested):
Hey @nikandfor! 

Primary issue to resolve here would be for the diffs to be presentable. Every PR gets deployed automatically to production. 

I just realized that I might've communicated this wrongly on Slack: 

<img width="509" alt="Screenshot 2022-10-07 at 13 45 08" src="https://user-images.githubusercontent.com/3893573/194556757-bd37aa3b-3ff0-4950-9ead-581ceb1a200e.png">

By "minimal" changes, I meant to say that we don't have to "refactor" code from the future stages right now, but we do need to make the changes required to make the diffs presentable. 

If it seems like a lot of wasted work to make the changes to future stages and you'd rather commit all refactored solutions at once, that's fine too. Upto you how you want to handle this, only constraint is that any diffs that land on the `master` branch should be presentable to users.

nikandfor commented:
In general in Go short variable names preferred: https://github.com/golang/go/wiki/CodeReviewComments#variable-names

So if we have a function on 40 lines and have variable like `objectFilePath` which we use only once in the next line this name is not common.

nikandfor commented:
@rohitpaulk  I've made some changes, could you give some feedback?

* `Mkdir` -> `MkdirAll` - it allows the program to finish successfully even if we already have some of the directories. Original git reinitializes repo with no error if it existed.
* Printed errors to stderr instead of stdout. This is a common practice for command line tools. This is especially useful when we use command output in a pipe and can lose an error message.
* split read blob command into a few smaller functions, each with fewer lines. Added Close calls, wrapped errors and made variable names a bit shorter.

Some parts could look a bit too complex like closing the file. But this little complexity is required for reliability.
```
    defer func() {
        e := file.Close()
        if err == nil && e != nil {
            err = fmt.Errorf("close file: %w", e)
        }
    }()
```

nikandfor commented:
At the end of stage 7, we will have a pretty big and complex app. And it will be reasonable to split it into smaller packages.

So choosing between smaller diffs (between stages) and simpler code in the first stages (it's not common to make multiple packages with a single function to solve a simple task), which is preferable?

rohitpaulk commented:
@nikandfor the points you mentioned above seem reasonable! 

I like how the rule around shorter variable names is phrased. Method receivers, and common things like readers can have short names, the more unusual something is the more descriptive the name should be. The current structure you have going looks good.

> At the end of stage 7, we will have a pretty big and complex app. And it will be reasonable to split it into smaller packages.

> So choosing between smaller diffs (between stages) and simpler code in the first stages (it's not common to make multiple packages with a single function to solve a simple task), which is preferable?

Simpler code in the first stages is preferable 🙂 The progressive "refactoring" in further stages is a good opportunity to educate people about code organization in Go, and in general feedback has been that users appreciate this.

Note that we won't need to do solutions for stage 7 of Git (it's pretty involved, and will likely be moved into a ["challenge extension"](https://app.codecrafters.io/vote/challenge-extension-ideas?course=git) later. Once we're done with solutions until stage 6 we can move on to other challenges. Not sure which at the moment, but I'll let you know once we're close to completion of stage 6.

rohitpaulk commented:
Just adding a checklist for myself to review the final diffs: 

- [ ] Stage 1
- [ ] Stage 2
- [ ] Stage 3
- [ ] Stage 4

rohitpaulk commented on solutions/go/read_blob/diff/cmd/mygit/cat_file_command.go.diff:
```diff
@@ -0,0 +1,97 @@
++
++	blobSha := args[2] // Assuming that args is ["cat-file", "-p", "hash"], just like os.Args
++
++	if len(blobSha) < 4 {
```
@nikandfor assuming this is supposed to check for the SHA length - should this be `len(blobSha) != 40`?

rohitpaulk commented on solutions/go/read_blob/diff/cmd/mygit/cat_file_command.go.diff:
```diff
@@ -0,0 +1,97 @@
++	blobSha := args[2] // Assuming that args is ["cat-file", "-p", "hash"], just like os.Args
++
++	if len(blobSha) < 4 {
++		return fmt.Errorf("bad object: %v", blobSha)
```
@nikandfor this isn't a "bad object", per-se - it's a bad SHA / object name. 

Here's what git says when a SHA is invalid: 

<img width="414" alt="Screenshot 2022-10-08 at 19 15 04" src="https://user-images.githubusercontent.com/3893573/194721739-9cb96bd0-d068-4c20-a65e-f230dc6bfaa8.png">

Can we do something similar here? Maybe `bad object name: %v`.

rohitpaulk commented on solutions/go/create_blob/diff/README.md.diff:
```diff
@@ -0,0 +1,59 @@
+ # Passing the first stage
+
+-The entry point for your Git implementation is in `cmd/mygit/main.go`. Study and
++The entry point for your Git implementation is in `app/main.go`. Study and
```
@nikandfor let's copy over the README from previous stages so that there's no diff here

nikandfor replied:
Yes, I wasn't cleaning all the stages after read blob yet.

rohitpaulk replied:
This specific case I'd fix now though - it can be confusing to users (given that the `app/main.go` file isn't even present). Refactoring code in further stages can of course be kept for future PRs.

rohitpaulk commented on solutions/go/create_blob/diff/cmd/mygit/main.go.diff:
```diff
@@ -0,0 +1,33 @@
+ 	case "cat-file":
+ 		err = catFileCmd(os.Args[1:])
++	case "hash-object":
++		err = HashObject(os.Args[2:])
```
@nikandfor you can tackle this in future PRs if you want, just thought I'd add a note here - the naming should be consistent among command functions, I'd expect this to be `hashObjectCmd`

nikandfor replied:
yes, I'm moving by small steps.

rohitpaulk reviewed (approved):
Good to merge once the remaining minor comments are addressed 🎉

nikandfor commented:
If I haven't missed some propagation it's done.

rohitpaulk commented:
@nikandfor looks good to me! Feel free to merge whenever you're ready 👍
